738 allow selection across cache boundaries

dstopp opened the pull request:
-on start of select range, pop up overlay with range start information
-allow selection of observations in range to span across cached data
-make sure range selection works in both directions

Fixes: #787, #738, #739

dstopp committed:
-store opusId, objNum and range action on start of range select in the infiniteScroll object
-when new data is retrieved and cache is flushed, no longer cancel range except on user request
-present start of range information in an overlay - still in progress to get it in a better location

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 738_allow_selection_across_cache_boundaries

dstopp committed:
-fixed the location of the range start overlay
-minor change to text of overlay

rfrenchseti reviewed (changes requested):
- I think the popup should distinguish between "Add range to cart starting at observation #x: opusid" and "Remove range from cart starting at observation #x: opusid". Just plain "Range select start" doesn't explain that it's a cart operation in progress. The wording given here matches that in the hamburger menu, and the format #x: yyy matches the hover text.
- When you start an add range and then drag the slider to a different cache region, the new thumbnail hamburger menus say "start add range" instead of "end add range".
- Now that we can add lots of observations, several bugs have been revealed that were probably there before but need to be fixed:
  - If adding the range fails because it exceeds the maximum number, the UI still selects the thumbnails even though the backend won't have added them to the cart. We need to wait for the addrange.json to return with a success code before changing the thumbnails. This will also solve the "no spinner" problem commented on around line 315 (which is not the same spinner problem mentioned below here).
  - When add.json or addrange.json returns with a failure code and an error dialog box pops up, the browse tab slider position is reset to 1 for no obvious reason.
  - When there are a lot of items in the cart and you switch to the cart tab, there is no spinner while it's loading the data until the very end (there's a long pause before the spinner starts)
- Another bug I noticed at the same time if you feel like fixing it here: If you pull down a hamburger menu and then move the slider, the hamburger menu stays present on the screen. The same is true of scrolling. Once this happens things get very confused if you start a range - the slider jumps.

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 738_allow_selection_across_cache_boundaries

dstopp committed:
-fixed hamburger menu remove/add text
-close menu on slider move

dstopp committed:
-close the hamburger menu on scroll of PS and mouse scroll.

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 738_allow_selection_across_cache_boundaries

dstopp committed:
-experimenting w/blocking clicks during cart operations

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 738_allow_selection_across_cache_boundaries

dstopp committed:
-do not move the slider to 1 on fail
-added spinner on cart operations
-disable clicking during while cart edit is working
-reorder highlight images/remove highlight of images after confirmation of edit complete

dstopp committed:
Removed the ability to edit cart until the recycle bin has been implemented.  this includes:
   - removing trash can icons from the tool and detail slide show view
   - removing the shift-click to start/end range functionality
   - removing the checkbox from the table view on the cart
Still allowed to remove all from cart.

rfrenchseti commented on opus/application/apps/ui/templates/ui/base.html:
```diff
@@ -350,6 +356,8 @@ <h3 class='op-header-text'></h3>
                 </div>
             </div>
         </div>
+		
```
Does this have the wrong indent level?

dstopp replied:
ok, looks like my editor is dropping tabs instead of spaces or something.  fixed.

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -631,17 +548,73 @@ var o_cart = {
+
+        let url = o_cart.getEditURL(opusIdRange, action);
+
+        // just so we don't have to continually checke for both addrange and add...
```
"checke"

dstopp replied:
done

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -631,17 +548,73 @@ var o_cart = {
+            }
+            opusIdRange = `${fromOpusId},${toOpusId}`;
+            let fromElem = $(`${tab} .op-gallery-view`).find(`[data-obs=${fromObsNum}]`);
+            let toELem = $(`${tab} .op-gallery-view`).find(`[data-obs=${toObsNum}]`);
```
ELem should not have a capital L

dstopp replied:
at least it was consistent :)  

done.

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -631,17 +548,73 @@ var o_cart = {
-        $.getJSON(url  + add_to_url + "&reqno=" + o_cart.lastRequestNo, function(statusData) {
-            // display modal for every error message returned from api call
+        // handle it as range
+        if (toOpusId != undefined) {
```
!==

dstopp replied:
done

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -631,17 +548,73 @@ var o_cart = {
+                length = $(`${tab} .op-thumbnail-container`).index(toELem);
+            }
+        } else {
+            // note - doing it this way handles the obs on the browse tab at the same time
```
What does "this way" mean?

dstopp replied:
removed comment, no longer applicable

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -631,17 +548,73 @@ var o_cart = {
+            let fromElem = o_browse.getGalleryElement(fromOpusId);
+            fromIndex = $(`${tab} .op-thumbnail-container`).index(fromElem);
+            action = (fromElem.hasClass("op-in-cart") ? "remove" : "add");
+        }
```
I think it would be more clear what's going on if we set "length = 1" here rather than using the default at the top of the function.

dstopp replied:
ok

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -631,17 +548,73 @@ var o_cart = {
+                fromIndex = $(`${tab} .op-thumbnail-container`).index(fromElem);
+            } else {
+                length = $(`${tab} .op-thumbnail-container`).index(toELem);
+            }
```
I think it would be more clear what's going on if we set "fromIndex = 0" here rather than using the default at the top of the function.

dstopp replied:
ok

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -631,17 +548,73 @@ var o_cart = {
+
+    toggleInCart: function(fromOpusId, toOpusId) {
+        let tab = opus.getViewTab();
+        let length = 1;
```
See comments below...could set length and fromIndex to undefined or null or something here to enforce their setting below and avoid potential bugs.

rfrenchseti reviewed (changes requested):
- When adding a range the little window as "Add range from cart" instead of "Add range to cart". The hamburger menu says this too.
- The "loading" spinner needs to have a delay on it just like the cart badge spinners do (see showCartCountSpinner). Right now every time you click on a trashcan/cart icon, you get a flicker of the big spinner even if it returns really fast.
- You need to look at updateBadges. I think its whole purpose is gone now that we don't need to add items to the cart one at a time when doing an addrange.

dstopp committed:
-fixed tab/space issue
-clarified variable setting per pull review request.

dstopp committed:
-added delay to '.loader' spinner to avoid flash when cart operation <or other things that utilize the .loader spinner>

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -54,6 +54,7 @@ var o_browse = {
     onRenderData: false,
     fading: false,  // used to prevent additional clicks until the fade animation complete
 
+    pageLoaderSpinnerTimer: null,   // used to apply a small delay to the loader on cart edit
```
This delay occurs in lots of places, not just on cart edit

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1128,16 +1131,16 @@ var o_browse = {
+        let rangeSelected = o_browse.isRangeSelectEnabled(tab);
+        let rangeText = "";
+        if (rangeSelected !== undefined) {
+            let addRemoveText = (rangeSelected === "removerange" ? "remove range from" : "add range from");
```
Here is where it should say "add range to"

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1128,16 +1131,16 @@ var o_browse = {
+            let addRemoveText = (rangeSelected === "removerange" ? "remove range from" : "add range from");
+            rangeText = `<i class='fas fa-sign-out-alt fa-rotate-180'></i>End ${addRemoveText} cart here`;
+        } else {
+            let addRemoveText = (inCart !== "in" ? "remove range from" : "add range from");
```
And here

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1178,7 +1181,17 @@ var o_browse = {
+        let obsNum = galleryElement.data("obs");
+        let action = (galleryElement.hasClass("op-in-cart") ? "removerange" : "addrange");
+        let actionText = (action === "removerange" ? "Remove" : "Add");
+        $(`${tab} .op-range-select-info-box`).html(`${actionText} range from cart starting at observation #${obsNum} ${opusId}  (ESC to cancel)`).addClass("op-range-select");
```
Also needs to change "from"/"to" here

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1191,13 +1204,36 @@ var o_browse = {
+
+    hidePageLoaderSpinner: function() {
+        if (o_browse.pageLoaderSpinnerTimer !== null) {
+            // This should always be true - we're just being careful
```
Actually I don't think this should always be true. For example, if you scroll such that the next page is being prefetched (spinner comes on), you can still click on a thumbnail to add it to the cart (also turns on spinner). Then when the page load is finished it turns off the spinner, and later when the cart operation returns it also tries to turn off the spinner, but the spinner is already off by then.

The right way to fix this is probably to reference count the on/off actions, since they should always be paired, and that way the spinner won't turn off until the last operation is complete. However, this sounds like a recipe for bugs, and it isn't a common occurrence, so I suggest we just remove this comment, perhaps put a new comment describing this issue, and forget about it for now.

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -526,92 +523,12 @@ var o_cart = {
-
     // action = add/remove/addrange/removerange/addall
-    editCart: function(opusId, action, updateBadges=true) {
+    getEditURL: function(opusId, action, updateBadges=true) {
```
Remove updateBadges

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -631,17 +548,74 @@ var o_cart = {
+        let action = null;
+        let elementArray = $(`${tab} .op-thumbnail-container`);
+        let opusIdRange = fromOpusId;
+        let updateBadges = true;
```
Remove updateBadges

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -650,18 +624,30 @@ var o_cart = {
+                    o_browse.updateCartIcon(opusId, status);
+                });
+                // we only update the cart badge result count from the latest request
+                if (updateBadges) {
```
Remove updateBadges

rfrenchseti reviewed (changes requested):
The following comments have not been addressed:

- When adding a range the little window as "Add range from cart" instead of "Add range to cart". The hamburger menu says this too.
- You need to look at updateBadges. I think its whole purpose is gone now that we don't need to add items to the cart one at a time when doing an addrange. Please remove it.

Also:

- When disabling clicking during cart operations, it only disables the gallery thumbnail area. You can still click on tab names. Is this the desired behavior?
- You can still Ctrl+click on an item in the cart tab (gallery or table) and have it add/remove from the cart. This needs to be disabled as well.

dstopp committed:
-changed the click-blocker to use the css pointer-events to none instead of overlay
-removed updateBadges
-generalized the disable of cart edit whilst on cart tab which should fix missing disable issues.
-added the pageload spinner to the cart gallery so the spinner shows while the cart details are collected before loadData can be called
-updated remove range from/add range to text per note
-updated comments re: pageLoaderSpinnerTimer
-added more detailed comments to hidePageLoaderSpinner function

dstopp committed:
-move the enable of the user interaction down a little bit more and move in update of global variables to further avoid potential race conditions
-fix the highlighted list of observations added to the cart <over cache boundaries it highlighted one less>

dstopp committed:
-keep the range select indicator on cache boundary changes until range select is done.

dstopp committed:
-fixed missing green box on table view over cache boundaries
